### PR TITLE
SPMD ForList support

### PR DESCRIPTION
**Pull request by zhangwen0411**

### CODE REVIEW

File: language/src/regent/alpha_convert.t [hunk 1]
```diff
@@ -167,6 +167,7 @@ local function alpha_convert_node(cx)
       node:is(ast.specialized.expr.ListPhaseBarriers) or
       node:is(ast.specialized.expr.ListInvert) or
       node:is(ast.specialized.expr.ListRange) or
+      node:is(ast.specialized.expr.ListIspace) or -- TODO: is this good?
```
Comment (zhangwen0411) [inline language/src/regent/alpha_convert.t]: Remove TODO.

File: language/src/regent/codegen.t [hunk 2]
```diff
@@ -1727,6 +1727,23 @@ function codegen.expr_field_access(cx, node)
...
+  elseif std.is_ispace(value_type) and field_name == "volume" then
+    local value = codegen.expr(cx, node.value):read(cx)
+    local expr_type = std.as_read(node.expr_type)
+
+    local volume = terralib.newsymbol(expr_type, "volume")
+    local actions = quote
+      [value.actions]
+      [emit_debuginfo(node)]
+
+      -- TODO(zhangwen): multiple domains?
```
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: std.assert no multiple domains.

File: language/src/regent/codegen.t [hunk 3]
```diff
@@ -3810,6 +3827,24 @@ function codegen.expr_cross_product_array(cx, node)
...
 end
 
+-- Returns Terra expression for subregion of partition of specified color.
+-- The color space can be either structured or unstructured, i.e. `color` can
+-- be an integer or of an index type.
+local function get_partition_subregion(cx, partition, color)
+  local color_type = color.type
+  assert(terralib.types.istype(color_type))
+
+  if std.validate_implicit_cast(color_type, int) then
```
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: Check partition:colors() dimension.  In unstructured case assert implicit cast.

File: language/src/regent/codegen.t [hunk 4]
```diff
@@ -3810,6 +3827,24 @@ function codegen.expr_cross_product_array(cx, node)
     expr_type)
 end
 
+-- Returns Terra expression for subregion of partition of specified color.
+-- The color space can be either structured or unstructured, i.e. `color` can
+-- be an integer or of an index type.
+local function get_partition_subregion(cx, partition, color)
```
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: Implicit-cast color to partition:colors().index_type, then call to_domain_point.

File: language/src/regent/std.t [hunk 5]
```diff
@@ -2158,6 +2181,40 @@ function std.index_type(base_type, displayname)
       })
   end
 
+  -- Element-wise min and max functions.
+  -- TODO(zhangwen): de-duplicate.
+  function st:e_min(expr1, expr2)
```
Comment (zhangwen0411) [inline language/src/regent/std.t]: Spell out "element".

File: language/src/regent/std.t [hunk 6]
```diff
@@ -2158,6 +2181,40 @@ function std.index_type(base_type, displayname)
...
+      in
+        (self { __ptr = [self.impl_type]
+          { [fields:map(function(field) return `(std.fmin(v1.[field], v2.[field])) end)] } })
+      end
+    else
+      return `(self { __ptr = [self.impl_type](std.fmin(expr1.__ptr, expr2.__ptr)) })
+    end
+  end
+
+  function st:e_max(expr1, expr2)
```
Comment (zhangwen0411) [inline language/src/regent/std.t]: Have a function that makes element-wise Terra functions.  st. ... = f(...).

File: language/src/regent/codegen.t [hunk 7]
```diff
@@ -4277,6 +4342,11 @@ function codegen.expr_list_invert(cx, node)
     [emit_debuginfo(node)]
   end
 
+  local color_type = rhs_type:partition():colors().index_type
+  if color_type:is_opaque() then  -- Treat `ptr` as `int1d`.
```
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: Why??

File: language/src/regent/codegen.t [hunk 8]
```diff
@@ -4262,6 +4295,38 @@ function codegen.expr_list_phase_barriers(cx, node)
...
+-- Returns Terra expression for color as an index_type (e.g. int2d).
+local function get_logical_region_color(cx, region, color_type)
+  assert(terralib.types.istype(color_type))
+  local domain_pt_expr =
+    `(c.legion_logical_region_get_color_domain_point([cx.runtime], region))
+  return color_type:from_domain_point(domain_pt_expr)
+end
+
+-- Returns Terra expression for offset of `point` in a rectangle.
+local function get_offset_in_rect(point, rect_size)
```
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: Make it not blow up for custom index types.  (iterate through fields)

File: language/src/regent/codegen.t [hunk 9]
```diff
@@ -4461,6 +4537,70 @@ function codegen.expr_list_range(cx, node)
...
+  local expr_type = std.as_read(node.expr_type)
+
+  local result = terralib.newsymbol(expr_type, "result") -- Resulting list.
+
+  -- Construct AST that populates the `result` list with indices from `ispace`.
+  --
+  -- Loop variable: `for p in ispace do`
+  local p_symbol = std.newsymbol(
+    -- TODO(zhangwen): how necessary is it to give a "bound"?  is this how you do it?
+    ispace_type.index_type(std.newsymbol(ispace_type)),
```
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: local bound
  if value:is(ast.typed.expr.ID) then
    bound = value.value
  else
    bound = std.newsymbol(value_type)
  end

File: language/src/regent/codegen.t [hunk 10]
```diff
@@ -4461,6 +4537,70 @@ function codegen.expr_list_range(cx, node)
...
+    "p")
+  -- Index in list: `result[i] = p; i += 1`.
+  local i = terralib.newsymbol(int, "i")
+
+  local loop_body = ast.typed.stat.Internal {
+    actions = quote
+      [expr_type:data(result)][ [i] ] = [p_symbol:getsymbol()];
```
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: Assert i is in bounds.

File: language/src/regent/codegen.t [hunk 11]
```diff
@@ -4461,6 +4537,70 @@ function codegen.expr_list_range(cx, node)
...
+    annotations = ast.default_annotations(),
+    span = node.span,
+  }
+
+  local actions = quote
+    [ispace.actions]
+    [emit_debuginfo(node)]
+
+    -- Compute size of resulting list.
+    -- TODO(zhangwen): multiple domains?
```
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: Assert.

File: language/src/regent/codegen.t [hunk 12]
```diff
@@ -5996,6 +6139,10 @@ local function cleanup_after(cx, block)
   return quote [result] end
 end
 
+function codegen.stat_internal(cx, node)
+  return node.actions
```
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: Assert that it is a quote.

File: language/src/regent/std.t [hunk 13]
```diff
@@ -2142,6 +2155,16 @@ function std.index_type(base_type, displayname)
     return [make_domain_point(self)]
   end
 
+  function st:from_domain_point(pt_expr)
```
Comment (zhangwen0411) [inline language/src/regent/std.t]: terra st.from_domain_point

File: runtime/legion/legion_c.cc [hunk 14]
```diff
@@ -1925,6 +1925,16 @@ legion_logical_region_get_color(legion_runtime_t runtime_,
   return runtime->get_logical_region_color(handle);
 }
 
+legion_domain_point_t
```
Comment (zhangwen0411) [inline runtime/legion/legion_c.cc]: Split into separate commit.

File: language/src/regent/codegen.t (hunk 7 above)
Comment (zhangwen0411) [inline language/src/regent/codegen.t]: `std.rect_type(color_type)` blows up if `color_type` is opaque.

**Comment by elliottslaughter** [comment]
Merged.
